[housekeeping] Use MSBuild for provisioning Android SDK, Appium

Redth opened the pull request:
This switches Appium and Android SDK to provision via MSBuild targets instead of cake/ps1 scripts and puts all of the version and api level/sdk package info into the Version.props to centralize the source of truth.

Redth committed:
Use MSBuild for provisioning

Appium and Android SDK to start

Redth committed:
We use Nexus 5X as the device type currently

rmarinho commented:
When using from the arcade approach it seems there are some issues. 

So the usage of calling the tools.csproj can be triggered by 

```
PS C:\repos\dotnet\maui> .\build.cmd -pack -verb:diag
 ```

rmarinho reviewed (changes requested):
When using from the arcade approach it seems there are some issues. 

So the usage of calling the tools.csproj can be triggered by 

```
PS C:\repos\dotnet\maui> .\build.cmd -pack -verb:diag
 ```

Redth committed:
Fix provision call / logs path

Redth committed:
Ignore exit codes

NPM may not return an ok exit code if there's no appium installed.  Similarly, we may get no driver list versions and a failed exit code when we query for appium drivers if none are installed.

Redth committed:
Fix call to target

Redth committed:
Provision JDK as well

Redth committed:
Provision from pipeline

Redth commented:
/azp run

Redth committed:
yaml fixes

Redth committed:
Yaml fix

Redth committed:
Add project to build invocation

Redth committed:
Restore dotnet tools first

Redth committed:
Restore tools after clear cache

Redth committed:
Ensure winget is installed first

Redth committed:
Download jdk msi and install it

Redth committed:
Different download location

Redth committed:
Property name for downloaded file

Redth committed:
Separate provisioning steps

PureWeen reviewed (changes requested):
rebase

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Log more jdk list info

Redth committed:
Update provision.yml

Redth committed:
Fix batching

Redth committed:
Log msiexec install

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Split up android provisioning steps

Redth committed:
Fix indentation

Redth committed:
Update tool with bugfix

Redth committed:
JDK: Try default installdir on windows

Redth committed:
Android: Option to install only default api levels

Redth committed:
Update provision.yml

rmarinho commented:
Make sure to rebase with latest changes the remove provisionator

rmarinho commented on eng/pipelines/common/provision.yml:
```diff
@@ -19,16 +22,37 @@ parameters:
-  #       SelfHealPowerShell: false
-  #       AUTH_TOKEN_GITHUB_COM: $(github--pat--vs-mobiletools-engineering-service2)
+  - ${{ if ne(parameters.poolName, 'Azure Pipelines') }}:
+    - template: agent-cleanser/v1.yml@yaml-templates
```
this can be kept disable I think

Redth committed:
Fix parameter passing

Redth committed:
MSBuild log diag

Redth committed:
Skip setting JAVA_HOME

Redth committed:
Compatible gradle versions for JDK 17

Redth committed:
Set JDK_HOME again

Redth committed:
Try to pick up jdk for gradle

Redth committed:
Fix type/var name

Redth committed:
Always look up android home

Redth committed:
Simplify locate of best JDK

Redth committed:
Cake runs with stdout

Redth committed:
Disable the cleaner as per comment

Redth committed:
Less verbose logging

Redth committed:
Update and comment out Tools entry point for now

Redth commented:
> When using from the arcade approach it seems there are some issues.
> 
> So the usage of calling the tools.csproj can be triggered by
> 
> ```
> PS C:\repos\dotnet\maui> .\build.cmd -pack -verb:diag
> ```

I commented out things here but have them updated to what they could be... not sure if we want to try and do that in this PR, but we can move that direction eventually?

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

rmarinho commented on src/Core/AndroidNative/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.6.1-all.zip
```
Why this downgrade?

Redth replied:
We're using JDK17...

Redth replied:
https://docs.gradle.org/current/userguide/compatibility.html suggests 7.3 but I had problems using 7.3 and compiling still so I went slightly higher... When I tried 7.6.x it complained about needing JDK21.

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Remove dup merge step

rmarinho commented on src/Provisioning/Provisioning.csproj:
```diff
@@ -0,0 +1,251 @@
+		<!-- Install any drivers not yet installed (if any) -->
+		<Exec Command="appium driver install %(AppiumDriversToInstall.Identity)@%(AppiumDriversToInstall.Version)" Condition=" '@(AppiumDriversToInstall)' != '' " />
+
+		<!-- Run doctor command just so we have some output -->
```
Can we add a Condition to the doctor , it caused issues like we saw on the agents where on macOS it hanged checking something with Xcode, and we can have that flag off to macOS  for now like I did in my last commit to main

Redth committed:
Add condition for skipping appium doctor

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

Redth commented:
/azp run

Redth committed:
Merge branch 'main' into dev/redth/provision-with-msbuild

mattleibow committed:
Merge remote-tracking branch 'origin/main' into dev/redth/provision-with-msbuild

# Conflicts:
#	eng/cake/dotnet.cake
#	eng/pipelines/common/provision.yml
#	eng/scripts/appium-install.ps1

mattleibow commented:
I have based my PR to make things work better and on linux and hopefully faster: https://github.com/dotnet/maui/pull/28065

rmarinho commented:
This was fixed by https://github.com/dotnet/maui/pull/28065
